handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -0,0 +1,1951 @@
+                    } catch (InvocationTargetException e) {
+                        throw new Error(e);
+                    }
+                }
+                if (SET_USE_CIPHER_SUITES_ORDER_METHOD != null && !isDestroyed()) {
+                    try {
+                        SET_USE_CIPHER_SUITES_ORDER_METHOD.invoke(sslParameters,
+                                (SSL.getOptions(ssl) & SSL.SSL_OP_CIPHER_SERVER_PREFERENCE) != 0);
+                    } catch (IllegalAccessException e) {
+                        throw new Error(e);
+                    } catch (InvocationTargetException e) {
+                        throw new Error(e);
+                    }
+                }
+            }
+        }
+        return sslParameters;
+    }
+
+    @Override
+    public final synchronized void setSSLParameters(SSLParameters sslParameters) {
+        super.setSSLParameters(sslParameters);
+
+        int version = PlatformDependent.javaVersion();
+        if (version >= 7) {
+            endPointIdentificationAlgorithm = sslParameters.getEndpointIdentificationAlgorithm();
+            algorithmConstraints = sslParameters.getAlgorithmConstraints();
+            if (version >= 8) {
+                if (SNI_HOSTNAME_CLASS != null && clientMode && !isDestroyed()) {
+                    assert GET_SERVER_NAMES_METHOD != null;
+                    assert GET_ASCII_NAME_METHOD != null;
+                    try {
+                        List<?> servernames = (List<?>) GET_SERVER_NAMES_METHOD.invoke(sslParameters);
+                        if (servernames != null) {
+                            for (Object serverName : servernames) {
+                                if (SNI_HOSTNAME_CLASS.isInstance(serverName)) {
+                                    SSL.setTlsExtHostName(ssl, (String) GET_ASCII_NAME_METHOD.invoke(serverName));
+                                } else {
+                                    throw new IllegalArgumentException("Only " + SNI_HOSTNAME_CLASS.getName()
+                                            + " instances are supported, but found: " +
+                                            serverName);
+                                }
+                            }
+                        }
+                        sniHostNames = servernames;
+                    } catch (IllegalAccessException e) {
+                        throw new Error(e);
+                    } catch (InvocationTargetException e) {
+                        throw new Error(e);
+                    }
+                }
+                if (GET_USE_CIPHER_SUITES_ORDER_METHOD != null && !isDestroyed()) {
+                    try {
+                        if ((Boolean) GET_USE_CIPHER_SUITES_ORDER_METHOD.invoke(sslParameters)) {
+                            SSL.setOptions(ssl, SSL.SSL_OP_CIPHER_SERVER_PREFERENCE);
+                        } else {
+                            SSL.clearOptions(ssl, SSL.SSL_OP_CIPHER_SERVER_PREFERENCE);
+             
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Inconsistent synchronization of io.netty.handler.ssl.ReferenceCountedOpenSslEngine.ssl; locked 88% of time [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIS2_INCONSISTENT_SYNC)